exporter/ststopologyexporter/internal/topology.go

```diff
@@ -0,0 +1,158 @@
+package internal
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type ComponentsCollection struct {
+	namespaces       map[string]*Component
+	services         []*Component
+	serviceInstances []*Component
+}
+
+func NewCollection() *ComponentsCollection {
+	return &ComponentsCollection{
+		make(map[string]*Component, 0),
+		make([]*Component, 0),
+		make([]*Component, 0),
+	}
+}
+
+func (c *ComponentsCollection) AddResource(attrs *pcommon.Map) bool {
+	serviceName, ok := attrs.Get("service.name")
+	if !ok {
+		return false
+	}
+	serviceNamespace, ok := attrs.Get("service.namespace")
+	if !ok {
+		return false
+	}
+	instanceId, ok := attrs.Get("service.instance.id")
+	var serviceInstanceId pcommon.Value
+	if !ok {
+		serviceInstanceId = serviceName
+	} else {
+		serviceInstanceId = instanceId
+	}
+
+	if _, ok := c.namespaces[serviceNamespace.AsString()]; !ok {
+		c.namespaces[serviceNamespace.AsString()] = &Component{
+			fmt.Sprintf("urn:opentelemetry:namespace/%s", serviceNamespace.AsString()),
+			ComponentType{
+				"namespace",
+			},
+			newData().
+				withLayer("urn:stackpack:common:layer:applications").
+				withEnvironment(attrs).
+				withName(attrs, "service.namespace"),
+		}
+	}
+	c.services = append(c.services, &Component{
+		fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s", serviceNamespace.AsString(), serviceName.AsString()),
+		ComponentType{
+			"service",
+		},
+		newData().
+			withLayer("urn:stackpack:common:layer:services").
+			withEnvironment(attrs).
+			withName(attrs, "service.name").
+			withVersion(attrs, "service.version").
+			withTags(attrs, "telemetry.sdk"),
+	})
+	c.serviceInstances = append(c.serviceInstances, &Component{
+		fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s:serviceInstance/%s", serviceNamespace.AsString(), serviceName.AsString(), serviceInstanceId.AsString()),
+		ComponentType{
+			"service_instance",
+		},
+		newData().
+			withLayer("urn:stackpack:common:layer:containers").
+			withEnvironment(attrs).
+			withName(attrs, "service.instance.id").
+			withVersion(attrs, "service.version").
+			withTags(attrs, "telemetry.sdk").
+			withTags(attrs, "telemetry.distro").
+			withProperties(attrs),
```

Comment:
These can just go into tags I think. We are very likely wanting to search by resource attributes (like by namespace)